File: ios/SmileIDExpoModule.swift
import ExpoModulesCore
import SmileID
import SwiftUI
import UIKit

public class SmileIDExpoModule: Module {

    // Each module class must implement the definition function. The definition consists of components
    // that describes the module's functionality and behavior.
    // See https://docs.expo.dev/modules/module-api for more details about available components.
    public func definition() -> ModuleDefinition {
        // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
        // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
        // The module will be accessible from `requireNativeModule('SmileIDExpo')` in JavaScript.
        Name("SmileIDExpo")

        // Defines a JavaScript function that always returns a Promise and whose native code
        // is by default dispatched on the different thread than the JavaScript runtime runs on.
       AsyncFunction("initialize") {
           (
               useSandBox: Bool,
               enableCrashReporting: Bool,
               config: SmileConfigRecord?,
               apiKey: String?
           ) async throws -> Void in
           
           if let apiKey = apiKey, let config = config {
               // Case 1: Initialize with API key and config
              SmileID.initialize(
                apiKey: apiKey,
                config: Config(
                    partnerId: config.partnerId,
                    authToken: config.authToken,
                    prodLambdaUrl: config.prodLambdaUrl,
                    testLambdaUrl: config.testLambdaUrl
                ),
                useSandbox: useSandBox
               )
           } else if let config = config {
               // Case 2: Initialize with just config
               SmileID.initialize(
                   config: Config(
                       partnerId: config.partnerId,
                       authToken: config.authToken,
                       prodLambdaUrl: config.prodLambdaUrl,
                       testLambdaUrl: config.testLambdaUrl
                   ),
                   useSandbox: useSandBox
               )
           } else {
               // Case 3: Basic initialization
               SmileID.initialize(useSandbox: useSandBox)
           }
       }


        // Document Verification View
        View(SmileIDDocumentVerificationView.self) {
            Events("onResult", "onError")
            
            Prop("config") { (
                view: SmileIDDocumentVerificationView,
                config: DocumentVerificationRequest
            ) in
                view.updateConfig(config)
            }
        }

        // Enhanced Document Verification View
        View(SmileIDDocumentVerificationEnhancedView.self) {
            Events("onResult", "onError")

            Prop("config") { (
                view: SmileIDDocumentVerificationEnhancedView,
                config: EnhancedDocumentVerificationRequest
            ) in
                view.updateConfig(config)
            }
        }

        // SmartSelfie Enrollment View
        View(SmileIDSmartSelfieEnrollmentView.self) {
            Events("onResult", "onError")


            Prop("config") { (
                view: SmileIDSmartSelfieEnrollmentView,
                config: SmartSelfieEnrollmentRequest
            ) in
                view.updateConfig(config)
            }
        }
        
        // Enhanced SmartSelfie Enrollment View
        View(SmileIDSmartSelfieEnrollmentEnhancedView.self) {
            Events("onResult", "onError")

            Prop("config") { (
                view: SmileIDSmartSelfieEnrollmentEnhancedView,
                config: SmartSelfieEnrollmentRequest
            ) in
                view.updateConfig(config)
            }
        }
        
        // SmartSelfie Authentication View
        View(SmileIDSmartSelfieAuthenticationView.self) {
            Events("onResult", "onError")
            
            Prop("config") { (
                view: SmileIDSmartSelfieAuthenticationView,
                config: SmartSelfieEnrollmentRequest
            ) in
                view.updateConfig(config)
            }
        }
        
        // Enhanced SmartSelfie Authentication View
        View(SmileIDSmartSelfieAuthenticationEnhancedView.self) {
            Events("onResult", "onError")
            
            Prop("config") { (
                view: SmileIDSmartSelfieAuthenticationEnhancedView,
                config: SmartSelfieEnrollmentRequest
            ) in
                view.updateConfig(config)
            }
        }
        
        // Biometric KYC View
        View(SmileIDBiometricKYCView.self) {
            Events("onResult", "onError")
            
            Prop("config") { (
                view: SmileIDBiometricKYCView,
                config: BiometricKYCRequest
            ) in
                view.updateConfig(config)
            }
        }
        
    }
}

File: android/src/main/java/com/smileidentity/react/expo/SmileIDExpoModule.kt
package com.smileidentity.react.expo

import com.smileidentity.SmileID
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import com.smileidentity.metadata.models.WrapperSdkName

class SmileIDExpoModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("SmileIDExpo")

        AsyncFunction("initialize") Coroutine { useSandBox: Boolean, enableCrashReporting: Boolean, config: SmileConfigRecord?, apiKey: String? ->
            val context = appContext.reactContext
                ?: throw IllegalStateException("Context is not available")

            withContext(Dispatchers.IO) {
                SmileID.setWrapperInfo(WrapperSdkName.ReactNative, "11.0.0")

                when {
                    // Case 1: Initialize with API key and config
                    apiKey != null && config != null -> {

                        val result = SmileID.initialize(
                            context = context,
                            config = config.toConfig(),
                            useSandbox = useSandBox,
                            enableCrashReporting = enableCrashReporting,
                            apiKey = apiKey
                        ).await()

                        result.getOrThrow()
                    }
                    // Case 2: Initialize with just config
                    config != null -> {
                        val result = SmileID.initialize(
                            context = context,
                            config = config.toConfig(),
                            useSandbox = useSandBox,
                            enableCrashReporting = enableCrashReporting,
                        ).await()

                        result.getOrThrow()
                    }
                    // Case 3: Basic initialization
                    else -> {
                        val result = SmileID.initialize(
                            context = context,
                            useSandbox = useSandBox,
                        ).await()
                        result.getOrThrow()
                    }
                }
            }
        }

        View(SmileIDDocumentVerificationView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDDocumentVerificationView, config: DocumentVerificationRecord ->
                view.updateConfig(config)
            }
        }

        View(SmileIDDocumentVerificationEnhancedView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDDocumentVerificationEnhancedView, config: EnhancedDocumentVerificationRecord ->
                view.updateConfig(config)
            }
        }

        View(SmileIDSmartSelfieEnrollmentView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDSmartSelfieEnrollmentView, config: SmartSelfieRecord ->
                view.updateConfig(config)
            }
        }

        View(SmileIDSmartSelfieEnrollmentEnhancedView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDSmartSelfieEnrollmentEnhancedView, config: SmartSelfieRecord ->
                view.updateConfig(config)
            }
        }

        View(SmileIDSmartSelfieAuthenticationView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDSmartSelfieAuthenticationView, config: SmartSelfieRecord ->
                view.updateConfig(config)
            }
        }

        View(SmileIDSmartSelfieAuthenticationEnhancedView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDSmartSelfieAuthenticationEnhancedView, config: SmartSelfieRecord ->
                view.updateConfig(config)
            }
        }

        View(SmileIDBiometricKYCView::class) {
            Events("onResult", "onError")
            Prop("params") { view: SmileIDBiometricKYCView, config: BiometricKYCRecord ->
                view.updateConfig(config)
            }
        }
    }
}
